hydra-test.el

```diff
@@ -31,80 +31,123 @@
-      (defun hydra-error/first-error nil
-        "Create a hydra with a \"M-g\" body and the heads:
+    (equal
+      (macroexpand
+        '(defhydra hydra-error (global-map "M-g")
+           "error"
+           ("h" first-error "first")
+           ("j" next-error "next")
+           ("k" previous-error "prev")
+           ("SPC" hydra-repeat "rep" :bind nil)))
+      '(progn
+         (set
+           (defvar hydra-error/keymap nil
+             "Keymap for hydra-error.")
+           (quote
+             (keymap
+               (32 . hydra-repeat)
+               (107 . hydra-error/previous-error)
+               (106 . hydra-error/next-error)
+               (104 . hydra-error/first-error)
+               (kp-subtract . hydra--negative-argument)
+               (kp-9 . hydra--digit-argument)
+               (kp-8 . hydra--digit-argument)
+               (kp-7 . hydra--digit-argument)
+               (kp-6 . hydra--digit-argument)
+               (kp-5 . hydra--digit-argument)
+               (kp-4 . hydra--digit-argument)
+               (kp-3 . hydra--digit-argument)
+               (kp-2 . hydra--digit-argument)
+               (kp-1 . hydra--digit-argument)
+               (kp-0 . hydra--digit-argument)
+               (57 . hydra--digit-argument)
+               (56 . hydra--digit-argument)
+               (55 . hydra--digit-argument)
+               (54 . hydra--digit-argument)
+               (53 . hydra--digit-argument)
+               (52 . hydra--digit-argument)
+               (51 . hydra--digit-argument)
+               (50 . hydra--digit-argument)
+               (49 . hydra--digit-argument)
+               (48 . hydra--digit-argument)
+               (45 . hydra--negative-argument)
+               (21 . hydra--universal-argument))))
+         (set
+           (defvar hydra-error/heads nil
+             "Heads for hydra-error.")
+           (quote
+             (("h"
+                first-error
+                "first"
+                :exit nil)
+               ("j"
+                 next-error
+                 "next"
+                 :exit nil)
+               ("k"
+                 previous-error
+                 "prev"
+                 :exit nil)
+               ("SPC"
+                 hydra-repeat
+                 "rep"
+                 :bind nil
+                 :exit nil))))
+         (set
+           (defvar hydra-error/hint nil
+             "Dynamic hint for hydra-error.")
+           (quote
+             (let ((format-statement (format
+                                       #("error: [h]: first, [j]: next, [k]: prev, [SPC]: rep."
+                                          8 9 (face hydra-face-red)
+                                          20 21 (face hydra-face-red)
+                                          31 32 (face hydra-face-red)
+                                          42 45 (face hydra-face-red)))))
+               (let ((result (eval format-statement)))
+                 (cl-loop
```

Comment:
This is what I initially referred to by «corner that had to be cut». Hydras may contain something like ``%20`foo``. Now consider this:

```
(with-current-buffer-window
 "baz"
 nil
 nil
 (insert
  (format
   (propertize "Magic number: %20s"
               'face
               '(:background "red"))
   42)))
```

Because of the padding added, `42` will not be propertized. This is why I came to believe that faces must be set at run-time, not when compiling the format string. Do you see another way to handle these cases?